core/unix/os.c

```diff
@@ -3725,6 +3725,18 @@ client_thread_run(void)
     byte *xsp;
     GET_STACK_PTR(xsp);
     void *crec = get_clone_record((reg_t)xsp);
+    /* i#2335: we support setup separate from start, and we want to allow a client
+     * to create a client thread during init, but we do not support that thread
+     * executing until the app has started (b/c we have no signal handlers in place).
+     */
+    /* i#3973: in addition to _executing_ a client thread before the
+     * app has started, if we even create the thread before
+     * dynamo_initialized is set, we will not copy tls blocks.  By
+     * waiting for the app to be started before dynamo_thread_init is
+     * called, we ensure this race condition can never happen, since
+     * dynamo_initialized will always be set before the app is started.
```

Comment:
Hmm, do we need to do this for Windows too?  It won't have this private loader TLS issue but it will run DR code in a separate thread before DR init is fully done which I think violates some assumptions.  For Windows we would move this wait from win32/os.c to win32/callback.s intercept_new_thread where it checks whether it's a client thread right before calling dynamo_thread_init.

If you do not want to deal with Windows you could bail on my suggestion and leave the issue open (or maybe file a new issue) for Windows, or if you want to put the suggested core/win32/ change in place and your new test passes on Appveyor we could then claim to have fully solved this related issue too for all platforms.

Reply:
I made the change for Windows too.

Unfortunately, my test change is causing a crash on Windows.  It was happening before I changed the Windows code, so I think it is the test itself since all the other changes were in unix code.

I don't have a Windows development machine handy, so it is more difficult for me to investigate that problem...

Reply:
OK maybe I can run the test on Windows in the next day or so.

Reply:
Here is the crash:

```
0:001> .cxr 0x00000277`7a50e800
rax=0000000000000004 rbx=ffffffffffffffff rcx=0000000000000000
rdx=0000000000000000 rsi=0000000000000000 rdi=000002777a496ac0
rip=00007ff725e3140c rsp=000002777a50ef20 rbp=00007ff725e31370
 r8=000000001536eb50  r9=0000000000000000 r10=0000000000000000
r11=0000000000000000 r12=000000f22b3ff8e0 r13=0000000000000000
r14=0000000000003838 r15=0000000000000000
iopl=0         nv up ei pl zr na po nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010246
client_thread_dll!thread_func+0x9c:
00007ff7`25e3140c 488b0cca        mov     rcx,qword ptr [rdx+rcx*8] ds:00000000`00000000=????????????????
0:001> kn
  *** Stack trace for last set context - .thread/.cxr resets it
 # Child-SP          RetAddr           Call Site
00 00000277`7a50ef20 00000000`1536ec98 client_thread_dll!thread_func+0x9c [d:\derek\dr\git\pr3976\suite\tests\client-interface\thread.dll.c @ 105] 
01 00000277`7a50efa0 00000000`1533eab2 dynamorio!client_thread_target+0x148 [d:\derek\dr\git\pr3976\core\win32\os.c @ 2801] 
02 00000277`7a50efe0 00000000`00000000 dynamorio!call_switch_stack+0x48 [D:\derek\dr\git\pr3976\build\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1667] 
0:001> U rip-f L10
client_thread_dll!thread_func+0x8d [d:\derek\dr\git\pr3976\suite\tests\client-interface\thread.dll.c @ 105]:
00007ff7`25e313fd 8b0dfd620300    mov     ecx,dword ptr [client_thread_dll!_tls_index (00007ff7`25e67700)]
00007ff7`25e31403 65488b142558000000 mov   rdx,qword ptr gs:[58h]
00007ff7`25e3140c 488b0cca        mov     rcx,qword ptr [rdx+rcx*8]
```
So it crashes accessing the "tls" variable, with NULL for its access method result:
```
    dr_fprintf(STDERR, "client thread is alive tls=%d\n", tls);
```

Reply:
Adding static TLS support to DR's private loader is #4030 and we need it for VS2017 anyway so I will try to add it.

Reply:
That's probably good.  I was hoping I'd have a chance to work on this over the holidays, but it never happened :man_shrugging: